makeflow/src/makeflow_wrapper_umbrella.c

```diff
@@ -0,0 +1,169 @@
+	debug(D_MAKEFLOW_RUN, "setting wrapper_umbrella->spec to %s\n", spec);
+}
+
+void makeflow_wrapper_umbrella_set_binary(struct makeflow_wrapper_umbrella *w, const char *binary) {
+	struct stat st;
+	if(lstat(binary, &st) == -1) {
+		fatal("lstat(`%s`) failed: %s\n", binary, strerror(errno));
+	}
+	if((st.st_mode & S_IFMT) != S_IFREG) {
+		fatal("the --umbrella-binary option of makeflow should binaryify a regular file\n");
+	}
+	w->binary = binary;
+	debug(D_MAKEFLOW_RUN, "setting wrapper_umbrella->binary to %s\n", binary);
+}
+
+void makeflow_wrapper_umbrella_preparation(struct makeflow_wrapper_umbrella *w, struct batch_queue *queue) {
+	// add umbrella_spec (if specified) and umbrella_binary (if specified) into the input file list of w->wrapper
+	if (!batch_queue_supports_feature(queue, "remote_rename")) {
+		if(w->spec) makeflow_wrapper_add_input_file(w->wrapper, w->spec);
+		if(w->binary) makeflow_wrapper_add_input_file(w->wrapper, w->binary);
+	} else {
+		if(w->spec) {
+			char *s = string_format("%s=%s", w->spec, path_basename(w->spec));
+			if(!s) fatal("string_format for umbrella spec failed: %s.\n", strerror(errno));
+			makeflow_wrapper_add_input_file(w->wrapper, s);
+			free(s);
+		}
+		if(w->binary) {
+			char *s = string_format("%s=%s", w->binary, path_basename(w->binary));
+			if(!s) fatal("string_format for umbrella binary failed: %s.\n", strerror(errno));
+			makeflow_wrapper_add_input_file(w->wrapper, s);
+			free(s);
+		}
+	}
+}
+
+char *create_umbrella_opt(struct batch_queue *queue, char *files, bool is_output) {
+	char *s = files;
+	size_t size;
+	char *result = NULL;
+	// construct the --output or --inputs option of umbrella based on files
+	while((size = strcspn(s, ",\0")) > 0) {
+		char *t;
+		s[size] = '\0';
+
+		if(!batch_queue_supports_feature(queue, "remote_rename")) {
+			t = s;
+		} else {
+			t = strchr(s, '=');
+			t++;
+		}
+
+		result = string_combine(result, t);
+		result = string_combine(result, "=");
+		result = string_combine(result, t);
+
+		if(is_output) result = string_combine(result, ":f,");
+		else result = string_combine(result, ",");
+
+		s[size] = ',';
+		s += size+1;
+	}
+
+	return result;
+}
+
+char *makeflow_wrap_umbrella(char *result, struct makeflow_wrapper_umbrella *w, struct batch_queue *queue, char *input_files, char *output_files) {
+	if(!w || !w->spec) return result;
+
+	char *umbrella_command = NULL;
+	char *umbrella_input_opt = NULL;
+	char *umbrella_output_opt = NULL;
+
+	debug(D_MAKEFLOW_RUN, "input_files: %s\n", input_files);
+	umbrella_input_opt = create_umbrella_opt(queue, input_files, false);
+	debug(D_MAKEFLOW_RUN, "umbrella input opt: %s\n", umbrella_input_opt);
+
+	debug(D_MAKEFLOW_RUN, "output_files: %s\n", output_files);
+	umbrella_output_opt = create_umbrella_opt(queue, output_files, true);
+	debug(D_MAKEFLOW_RUN, "umbrella output opt: %s\n", umbrella_output_opt);
+
+	// construct umbrella_command
+	if(!batch_queue_supports_feature(queue, "remote_rename")) {
+		if(!w->binary) {
```

Comment:
As with the above comment, this is allowable with the knowledge that the user should be warned that we are assuming the executable exists at the work site. Disregard my previous comments on this.